<!DOCTYPE html>
<html>
  <head>
    <title>RNA Dotplots, McCaskill and Locality</title>
    <meta charset="utf-8">
    <style>
       @import url(https://fonts.googleapis.com/css?family=Droid+Serif);
      @import url(https://fonts.googleapis.com/css?family=Yanone+Kaffeesatz);
      @import url(https://fonts.googleapis.com/css?family=Ubuntu+Mono:400,700,400italic);

      body {
        font-family: 'Droid Serif';
      }
      h1, h2, h3 {
        font-family: 'Yanone Kaffeesatz';
        font-weight: 400;
        margin-bottom: 0;
      }
      .remark-slide-content h1 { font-size: 3em; }
      .remark-slide-content h2 { font-size: 2em; }
      .remark-slide-content h3 { font-size: 1.6em; }
      .footnote {
        position: absolute;
        bottom: 3em;
      }
      li p { line-height: 1.25em; }
      .red { color: #fa0000; }
      .gray { color: #777; }
      .large { font-size: 2em; }
      .small { font-size: 0.5em; }
      a, a > code {
        color: rgb(249, 38, 114);
        text-decoration: none;
      }
      code {
        background: #e7e8e2;
        border-radius: 5px;
      }
      .remark-code, .remark-inline-code { font-family: 'Ubuntu Mono'; }
      .remark-code-line-highlighted     { background-color: #373832; }
      .pull-left {
        float: left;
        width: 47%;
      }
      .pull-right {
        float: right;
        width: 47%;
      }
      .pull-right ~ p {
        clear: both;
      }
      #slideshow .slide .content code {
        font-size: 0.8em;
      }
      #slideshow .slide .content pre code {
        font-size: 0.9em;
        padding: 15px;
      }
      .inverse {
        background: #272822;
        color: #777872;
        text-shadow: 0 0 20px #333;
      }
      .inverse h1, .inverse h2 {
        color: #f3f3f3;
        line-height: 0.8em;
      }

      /* Slide-specific styling */
      #slide-inverse .footnote {
        bottom: 12px;
        left: 20px;
      }
      #slide-how .slides {
        font-size: 0.9em;
        position: absolute;
        top:  151px;
        right: 140px;
      }
      #slide-how .slides h3 {
        margin-top: 0.2em;
      }
      #slide-how .slides .first, #slide-how .slides .second {
        padding: 1px 20px;
        height: 90px;
        width: 120px;
        -moz-box-shadow: 0 0 10px #777;
        -webkit-box-shadow: 0 0 10px #777;
        box-shadow: 0 0 10px #777;
      }
      #slide-how .slides .first {
        background: #fff;
        position: absolute;
        top: 20%;
        left: 20%;
        z-index: 1;
      }
      #slide-how .slides .second {
        position: relative;
        background: #fff;
        z-index: 0;
      }

      /* Two-column layout */
      .left-column {
        color: #777;
        width: 20%;
        height: 92%;
        float: left;
      }
        .left-column h2:last-of-type, .left-column h3:last-child {
          color: #000;
        }
      .right-column {
        width: 75%;
        float: right;
        padding-top: 1em;
      }

           /* Two-column layout */
      .my-left-column {

        width: 70%;
        float: left;
      }
       
      .my-right-column {
        color: #888;
        width: 15%;
        float: right;
        height: 92%;

        padding-top: 1em;
      }
       .my-right-column h2:last-of-type, .my-right-column h3:last-child {
          color: #001;
        }
    </style>
  </head>
  <body>
    <textarea id="source">
name: inverse
layout: true
class: center, middle, inverse
--- 

---
layout: false

.center[

## RNA Dotplots, 
## McCaskill 
### and the curse of Locality
]
---
layout: false

.center[Probability of an structure in the ensemble]
.center[![](prob-vis.svg)]
- BW:
     - Boltzmann Weight
     - Exponential function  => exponential scale behaviors!

- Z:  
     - Partition function
     - Sum of the Boltzmann weights for the entire ensemble

---

layout: false
.my-left-column[.center[##McCaskill, 1990, ]

  - For a given sequence, efficient methods for: 
  1. **partition function (Z)**  
    - Z(i,j) 
    - For all sub-sequences
        
  2. **probability of an individual base-pair** in ensemble 
      - p(i,j)   
      - For all possible pairs

  3. Visualizing all base-pair probabilities as **dot plot**
      - Area(i,j) = p(i,j) . Unit-Area
  
]
.my-right-column[
<img align="top" src="john_s_maccaskill_young.jpeg"  width="70" > 

 Z(1,72)=-25.45 kcal/mol

##### p(3,68)=0.9

![](X70810.2.svg)
]

---

layout: false
.my-left-column[.center[##McCaskill, 1990, ]
###Based on Turner energy model
###No heuristics and approximation in the computations!
]
.my-right-column[
<img align="top" src="john_s_maccaskill_young.jpeg"  width="70" > 

 Z(1,72)=-25.45 kcal/mol

##### p(3,68)=0.9

![](X70810.2.svg)
]

---

###.center[Dot plot, 1]
###Advantages:
1. An excellent overview of **high** probable base-pairs

2. Great help to detect the **second** probable structure. 
   - Ribo-switch/bistable RNAs for example
---
###.center[Dot plot, 2]
###Advantages:
1. An excellent overview of **high** probable base-pairs

2. Great help to detect the **second** probable structure. 
   - Ribo-switch/bistable RNAs for example

3. .gray["Integration Test"] <img align="top" src="emoji.jpg"  width="22" >
   - .gray[For the new comers in the field of RNA-bioinf ]

  .center[![](X70810.2-annotated.svg)]

---
 ### Dot plot: The evolution

.small[.footnote[*McCaskill's picture taken years are not exact :)]]

.center[![](dotplot-evolution.svg)]

*(CHURKIN et al., 2013) proposes an alternative for bistable RNAs

*RNAstruct provides a confusing colored dot-plot

---

###.center[Plotting medium has fixed size ]
- Right plot is only zoomed out, exact same probabilities as left
- The Human perception mainly ignores the small dots.

![](X70810.2_medium.svg)
![](X70810.2_scaling_problem.svg)

---
###.center[How low are the low probabilities? ]
.center[![](X70810.2_scaling_problem.svg)]

---
.center[![](sample_trna2_bp_probs.svg)]

---
.center[![](sample_trna2_bp_probs_inverse.svg)]

---
### .center[Other limitations ]

- Basepair probabilities are independent of each other.

- Therefor no explicit relationship between cells can be encoded!

- No one knows which stems exist together.
    - Exception is for the case of crossing base-pairs.

.center[![](X70810.2-annotated.svg)]

---
layout: false

.center[

### .gray[RNA Dotplots, ]
### .gray[McCaskill ]
## and the curse of Locality
]

---
.center[Let's play a game with this cute tRNA.]

- Extending from both ends
- Di-Nucleotide shuffled genomic context
- tRNA position is according to a normal distribution
- On average close to the center of the context 
.center[![](X70810.2_extend.svg)]


---
.center[Probability of the base-pair, by global folding]
- Each time the context is re-shuffled and re-sampled
![](tRNA-extend.svg)

---
.center[Probability of the selected base-pair, by global folding]
- Each time the context is re-shuffled and re-sampled
![](tRNA-Log--extend.svg)

---
<!-- .small[.footnote[*And always hides]] -->
Well this is the classic local folding problem
 
1. Lfold (2004): 
    * Ignore any possible base-pair longer than L.
    
.center[![](Lfold.svg)]

2. RNAplFold(2006): 
    * A follow-up: Enough to believe Lfold was not good enough  <img align="top" src="emoji.jpg"  width="22" >
    * Windowing:
        * Pragmatic
        * Usually works 
        * Can be found in the pocket of any bioinformatician

  

---
We have concisely showed that: (Sita, Daniel, Mathias, et al. 2012)

- Parameterization is very crucial to have accurate RNAplFold
 
####Tip: 

*Strong parameter dependency*:
- Usually a good sign that a tool does not work as supposed to do so.

---
.center[The second game: Splitting]
.center[![](X70810.2_split.svg)]

---
.center[Probability of the selected base-pair, by global folding]
- Each time the context is re-shuffled and re-sampled
![](tRNA-split.svg)

---
    </textarea>
    <script src="https://gnab.github.io/remark/downloads/remark-latest.min.js">
    </script>
    <script>
      var slideshow = remark.create();
    </script>
  </body>
</html>